### ISSUE TITLE
Slow performance iterating over a file

### ISSUE THREAD

**Issue by lpil** [richard-viney/file_streams#4]
Hello!

Theo was doing some benchmarking and discovered that this library is very slow, 12 times slower than NodeJS.

https://x.com/t3dotgg/status/1798234716312887448

Seeing as this is likely a library you will want to use in performance critical parts of your application it would be good to optimise this.

It looks like raw mode isn't being used- that is often a good place to start. Perhaps it should _always_ use raw mode, rather than expecting the programmer to know about tuning file IO performance on Erlang?

Thanks,
Louis

**Comment by richard-viney** [comment]
Hey Louis!

That's useful feedback, I've not done any performance testing to date.

Raw mode is certainly faster according to the docs, and was turned on for binary streams in v0.4.4, then turned off in v0.5 because it means ["the functions in the `io` module cannot be used"](https://www.erlang.org/doc/apps/kernel/file.html#open/2), and this library uses `io` functions to support non-Latin-1 (ISO 8859-1) text encodings for text file streams (UTF-8, UTF-16/32 LE/BE, etc.)

Raw mode should likely be re-enabled for binary file streams, but that wouldn't help Theo's situation as he wants to read UTF-8 lines, which AFAICT means using [`io.get_line`](https://www.erlang.org/doc/apps/stdlib/io#get_line/2) as [`file.read_line`](https://www.erlang.org/doc/apps/kernel/file.html#read_line/1) doesn't do UTF-8, which means raw mode can't be used.

Maybe better performance could be achieved by sticking to binary/raw streams and handling text encoding concerns somewhere other than the `io` module .. ?

Initial testing results:

- The current v0.5 of this library takes over 15 minutes to read and count a billion lines (I stopped timing at 15 minutes).
- Using raw mode and [`file.read_line`] reduced this to ~9 minutes, with the caveat that it doesn't work correctly with arbitrary UTF-8 text as detailed above.
- A simple Rust implementation streaming in and counting the lines took ~53 seconds.

This was on a MacBook Pro M2 with 1TB SSD.

Would also be worth testing with some different read-ahead buffer sizes as well to see what impact that has, but yes it's currently over an order of magnitude slower than Rust.

I should have time to dig a bit deeper tomorrow.

**Comment by richard-viney** [comment]
[v0.5.1](https://github.com/richard-viney/file_streams/releases/tag/v0.5.1) has been released that adds raw mode and enables it by default when opening binary file streams with `file_stream.open_read()` and `file-stream.open_write()`.

Raw mode can't currently be used on text file streams for the reasons detailed above, so this change doesn't help the original performance issue reported here.

**Comment by richard-viney** [comment]
[v0.6.0](https://github.com/richard-viney/file_streams/releases/tag/v0.6.0) fixes this!

tl;dr: Use `file_streams.open_read()`, and only reach for `file_streams.open_read_text()` if you need to read a non-UTF-8 text encoding.

---

Refer to the latest README for more detail about text encodings.

The original, slow path for reading and writing encoded text is still available, and is required for text in a non-UTF-8 encoding, but the most common case of reading lines of UTF-8 text is now ~10x faster.

On a one billion line text file from the 1BRC my timings are:

- Rust using `io::BufReader`: 0:53
- Node.js using `readline`: 2:05
- Gleam using `file_streams`: 2:25

The Gleam number goes down further to 1:41 if validation of the read bytes as UTF-8 is skipped. Gleam strings must be valid UTF-8 so this can't be skipped unless the input is known to be valid UTF-8. An option could be added for this if the extra performance boost is desired.
